proxy/GcsConnection.go

```diff
@@ -0,0 +1,300 @@
+	if err != nil {
+		log.Printf("ResponseMessage: Error writing header: %v", err)
+		return 0, err
+	}
+	copy(p, headerBuf)
+	n += len(headerBuf)
+
+	if resp.secondHeader != nil {
+		nSecond, err := resp.secondHeader.Read(p[n:])
+		n += nSecond
+		if err != nil && err != io.EOF {
+			log.Printf("ResponseMessage: Error reading secondHeader: %v", err)
+			return n, err
+		}
+		if err == io.EOF && resp.data != nil {
+			nData, err := resp.data.Read(p[n:])
+			n += nData
+			return n, err
+		}
+	}
+	if resp.data != nil {
+		nData, err := resp.data.Read(p[n:])
+		n += nData
+		return n, err
+	}
+	log.Println("ResponseMessage: Finished Read with EOF")
+	return n, io.EOF
+}
+
+type ConnectionGroup struct {
+	messages       chan *RequestMessage
+	responseMap    map[uint32]chan *ResponseMessage
+	responseMapMux sync.Mutex
+	address        string
+	ctx            context.Context
+	nConnections   int
+}
+
+func NewConnectionGroup(bufferSize int, address string, ctx context.Context, nConnections int) *ConnectionGroup {
+	log.Printf("Creating ConnectionGroup with bufferSize=%d, address=%s, nConnections=%d", bufferSize, address, nConnections)
+	cg := &ConnectionGroup{
+		messages:     make(chan *RequestMessage, bufferSize),
+		responseMap:  make(map[uint32]chan *ResponseMessage),
+		address:      address,
+		ctx:          ctx,
+		nConnections: nConnections,
+	}
+
+	for i := 0; i < nConnections; i++ {
+		go func(i int) {
+			log.Printf("Starting connection goroutine %d", i)
+			if err := cg.handleConnection(); err != nil {
+				log.Printf("Connection goroutine %d error: %v", i, err)
+			}
+		}(i)
+	}
+
+	return cg
+}
+
+func (cg *ConnectionGroup) createConnection() (net.Conn, error) {
+	log.Printf("Creating connection to %s", cg.address)
+	return net.Dial("tcp", cg.address)
+}
+
+func (cg *ConnectionGroup) readFromConn(conn net.Conn, readErrCh chan<- error) {
+	log.Println("Starting readFromConn")
+	for {
+		select {
+		case <-cg.ctx.Done():
+			log.Println("readFromConn: Context canceled")
+			readErrCh <- cg.ctx.Err()
+			return
+		default:
+			var resp ResponseMessage
+			if err := binary.Read(conn, binary.LittleEndian, &resp.header); err != nil {
+				log.Printf("readFromConn: Error reading header: %v", err)
+				readErrCh <- err
+				return
+			}
+
+			cg.responseMapMux.Lock()
+			if ch, exists := cg.responseMap[resp.header.RequestUid]; exists {
+				log.Printf("readFromConn: Sending response for RequestUid=%d", resp.header.RequestUid)
+				ch <- &resp
+				close(ch)
+				delete(cg.responseMap, resp.header.RequestUid)
+			}
+			cg.responseMapMux.Unlock()
+		}
+	}
+}
+
+func (cg *ConnectionGroup) writeToConn(conn net.Conn, writeErrCh chan<- error) {
+	log.Println("Starting writeToConn")
+	for {
+		select {
+		case req := <-cg.messages:
+			log.Printf("writeToConn: Sending request for RequestUid=%d", req.header.RequestUid)
+			if _, err := io.Copy(conn, req); err != nil {
+				log.Printf("writeToConn: Error writing request: %v", err)
+				writeErrCh <- err
```

Comment:
This notifies only `handleConnection()` about an error. How do you fail the `req`?